Adapted Phase2 Tracker Digitizer code to C++11 standard

suchandradutta opened the pull request:
#### PR description:
overall code optimization which includes 
- Optimization of loops over simhits with std::copy_if
- Use of std::make_unique (C++14)
- STL loop improvement
- Clean-up of unnecessary header files, temporary variables, superfluous code
- Style uniformity

#### PR validation:
Digis validated using FourFourMuon (Pt: 10-100 GeV) + 200 PU events in CMSSW_11_1_X
( workflow number 20610.0 )  against the released version

Details can be found in the presentation : https://indico.cern.ch/event/838963/contributions/3679018/attachments/1964986/3267155/Digitizer_Code_Review_19Dec19.pdf


Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)
@subirsarkar @emiglior

Commit:
Code Optimization : loop using std::copy_if, Use std::make_unique, STL loop improvement, Clean-up, Style uniformity

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13232

- This PR adds an extra 36KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13232/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13232/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13232/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13232/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

kpedro88 commented:
@suchandradutta please apply the code-checks and code-format, otherwise the PR tests will not start.
Also, please update the PR title to make it clear that it refers to the Phase 2 tracker digitization code.

Commit:
code-checks and code-format are performed

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13242

- This PR adds an extra 36KB to repository

cmsbuild commented:
A new Pull Request was created by  @suchandradutta (Suchandra Dutta) for master.

It involves the following packages:

SimTracker/SiPhase2Digitizer


@cmsbuild, @civanch, @kpedro88, @mdhildreth can you please review it and eventually sign? Thanks.
@makortel, @GiacomoSguazzoni, @rovere, @VinInn, @ebrondol, @threus, @dgulhan this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: bcf49c7f3ce120ca1d768b97aa6f46f176581b6c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d43235/4083/summary.html
CMSSW: CMSSW_11_1_X_2019-12-20-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d43235/4083/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2813777
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2813434
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

civanch commented:
+1

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc:
```diff
@@ -116,20 +97,20 @@ namespace cms {
           const Phase2TrackerGeomDetUnit* pixdet = dynamic_cast<const Phase2TrackerGeomDetUnit*>(det_u);
           assert(pixdet);
-          detectorUnits_.insert(std::make_pair(detId_raw, pixdet));
+          detectorUnits_.insert({rawId, pixdet});
```
better to use `emplace(rawId,pixdet)`

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc:
```diff
@@ -323,15 +306,14 @@ namespace cms {
-    std::unique_ptr<edm::DetSetVector<PixelDigi> > output(new edm::DetSetVector<PixelDigi>(digiVector));
-    std::unique_ptr<edm::DetSetVector<PixelDigiSimLink> > outputlink(
-        new edm::DetSetVector<PixelDigiSimLink>(digiLinkVector));
+    std::unique_ptr<edm::DetSetVector<PixelDigi> > output = std::make_unique<edm::DetSetVector<PixelDigi> >(digiVector);
```
to reduce type repetition, you can write these as:
```
auto output = std::make_unique<edm::DetSetVector<PixelDigi>>(digiVector);
```
etc.

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc:
```diff
@@ -340,34 +322,36 @@ namespace {
     // coded channel stays the same (so that it can then be decoded
     // with Phase2TrackerDigi in stage2).
-    std::pair<int, int> ip = PixelDigi::channelToPixel(channel);
+    auto ip = PixelDigi::channelToPixel(channel);
```
`const auto&` (also below)

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:
```diff
@@ -26,36 +22,29 @@
 
+namespace {
+  // Mass in MeV
+  constexpr double m_pion = 139.571;
```
we may want to consider a common source within CMSSW for these mass values (other values are used elsewhere in some cases)

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:
```diff
@@ -345,69 +321,67 @@ void Phase2TrackerDigitizerAlgorithm::drift(const PSimHit& hit,
-    SegX = ionization_points[i].x();
-    SegY = ionization_points[i].y();
-    SegZ = ionization_points[i].z();
+  //float Sigma_x = 1.;  // Charge spread
```
delete commented-out code

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:
```diff
@@ -345,69 +321,67 @@ void Phase2TrackerDigitizerAlgorithm::drift(const PSimHit& hit,
 
     // pseudoRadDamage
-    if (pseudoRadDamage >= 0.001) {
+    if (pseudoRadDamage_ >= 0.001) {
```
magic number 0.001 should be named constant

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:
```diff
@@ -495,82 +458,78 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(
       }
       float TotalIntegrationRange = UpperBound - LowerBound;  // get strip
-      x[ix] = TotalIntegrationRange;                          // save strip integral
+      x.insert({ix, TotalIntegrationRange});                  // save strip integral
```
use emplace

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:
```diff
@@ -495,82 +458,78 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(
 
       float TotalIntegrationRange = UpperBound - LowerBound;
-      y[iy] = TotalIntegrationRange;  // save strip integral
+      y.insert({iy, TotalIntegrationRange});  // save strip integral
```
use emplace

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:
```diff
@@ -889,20 +833,18 @@ void Phase2TrackerDigitizerAlgorithm::module_killing_DB(uint32_t detID) {
-      if (SiPixelBadModule_->IsRocBad(detID, j) == true) {
-        std::vector<CablingPathToDetUnit> path = map_.product()->pathToDetUnit(detID);
+    std::vector<GlobalPixel> badrocpositions;
+    for (size_t j = 0; j < 16; j++) {
```
magic number 16 should be named constant

kpedro88 reviewed (commented):
Thanks for these useful code improvements. In addition to the minor review comments, I have a few general comments:
* PSSDigitizerAlgorithm and PSPDigitizerAlgorithm appear to be the exact same code, just with a different name. I recommend making a generic PSDigitizerAlgorithm, and if you want to specify the PSS vs. PSP name for debugging etc., make that an input parameter of the algorithm.
* In general, we recommend that plugins just have a .cc file, with no separate .h file (since plugin code is not supposed to be included in other code).

suchandradutta committed:
Updating the package following comments in the PR (#28664)

cmsbuild commented:
The code-checks are being triggered in jenkins.

suchandradutta committed:
Updating the package following comments in the PR (#28664)

suchandradutta committed:
Updating the package following comments in the PR (#28664)

cmsbuild commented:
The code-checks are being triggered in jenkins.

suchandradutta committed:
Updating the package following comments in the PR (#28664)

suchandradutta committed:
Updating the package following comments in the PR (#28664)

cmsbuild commented:
The code-checks are being triggered in jenkins.

emiglior commented:
@skinnari FYI

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13365

- This PR adds an extra 36KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13365/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13365/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

suchandradutta commented:
Incorporated suggestion for minor changes. 

About the general comments
  

-   **PSSDigitizerAlgorithm and PSPDigitizerAlgorithm appear to be the exact same code, just with a different name. I recommend making a generic PSDigitizerAlgorithm, and if you want to specify the PSS vs. PSP name for debugging etc., make that an input parameter of the algorithm.**

==> We expect different implementations following individual signal shape parametrization of PSP and PSS algorithms. The noise treatment may also differ. Once the picture becomes clear we shall take proper action.

-  **In general, we recommend that plugins just have a .cc file, with no separate .h file (since plugin code is not supposed to be included in other code).** 

Header files for Individual algorithms are merged in the .cc and removed. plugins/Phase2TrackerDigitizer.h is used by plugins/PreMixingPhase2TrackerWorker.cc. Similarly, plugins/Phase2TrackerDigitizerAlgorithm.h is used by individual algorithms.

suchandradutta committed:
"code-check" and "code-format" done

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13366

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #28664 was updated. @cmsbuild, @civanch, @kpedro88, @mdhildreth can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

kpedro88 commented on SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc:
```diff
@@ -345,69 +321,65 @@ void Phase2TrackerDigitizerAlgorithm::drift(const PSimHit& hit,
 
     // pseudoRadDamage
-    if (pseudoRadDamage >= 0.001) {
+    if (pseudoRadDamage_) {
```
Is this change intentional? The behavior is different from the previous version of this line.

cmsbuild commented:
-1

Tested at: 878fd458d1631a6ac809db0647242e5d76bdd641

CMSSW: CMSSW_11_1_X_2020-01-14-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d43235/4231/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found compilation error when building: 

<pre>>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-01-14-2300/src/SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.cc 
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-01-14-2300/src/SimTracker/SiPhase2Digitizer/plugins/PreMixingPhase2TrackerWorker.cc 
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-01-14-2300/src/SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.cc 
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-01-14-2300/src/SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc 
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-01-14-2300/src/SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc:21:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-01-14-2300/poison/SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.h:1:2: error: #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
 #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
  ^~~~~
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-01-14-2300/src/SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc:22:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-01-14-2300/poison/SimTracker/SiPhase2Digitizer/plugins/PSSDigitizerAlgorithm.h:1:2: error: #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
 #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

suchandradutta committed:
Restored header files for  individual algorithms 

Realised that the header files for  individual algorithms are needed. Putting them back. If required we can move them to interface(.h) and src(.cc) areas.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28664/13370

- This PR adds an extra 40KB to repository

- Found files with invalid states:
  - SimTracker/SiPhase2Digitizer/plugins/PSPDigitizerAlgorithm.h:
    - Modified: b794f0d2a1cd03bfd508aa83a7bdb9e61114167e, bcf49c7f3ce120ca1d768b97aa6f46f176581b6c
    - Deleted: 3e82bb89fca6f3df758015a66541607756983824
    - Added: 91abd9cf8607b45edd51f0b391ded5b4c214f973
  - SimTracker/SiPhase2Digitizer/plugins/PixelDigitizerAlgorithm.h:
    - Modified: b794f0d2a1cd03bfd508aa83a7bdb9e61114167e, bcf49c7f3ce120ca1d768b97aa6f46f176581b6c
    - Deleted: e6e4960e79fb4f639f4568fa38c5bea93379ca97
    - Added: 91abd9cf8607b45edd51f0b391ded5b4c214f973
  - SimTracker/SiPhase2Digitizer/plugins/SSDigitizerAlgorithm.h:
    - Modified: b794f0d2a1cd03bfd508aa83a7bdb9e61114167e, bcf49c7f3ce120ca1d768b97aa6f46f176581b6c
    - Deleted: 006cf468b6cff14e1d6d5d4549fc7443855400b0
    - Added: 91abd9cf8607b45edd51f0b391ded5b4c214f973
  - SimTracker/SiPhase2Digitizer/plugins/PSSDigitizerAlgorithm.h:
    - Modified: b794f0d2a1cd03bfd508aa83a7bdb9e61114167e, bcf49c7f3ce120ca1d768b97aa6f46f176581b6c
    - Deleted: dbf5bac9b2ede00e709eab5a0bc64f08862dfad4
    - Added: 91abd9cf8607b45edd51f0b391ded5b4c214f973

cmsbuild commented:
Pull request #28664 was updated. @cmsbuild, @civanch, @kpedro88, @mdhildreth can you please check and sign again.

suchandradutta commented:
Prior to the last commit I suspect I had an un-cleaned build  and the compilation error did not show up! Sorry for that!

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 91abd9cf8607b45edd51f0b391ded5b4c214f973
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d43235/4232/summary.html
CMSSW: CMSSW_11_1_X_2020-01-14-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d43235/4232/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2711715
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2711368
 - DQMHistoTests: Total skipped: 346
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

silviodonato commented:
CMSSW_11_1_X will be used to run a rereco of GEN-SIM-DIGI-RAW produced with CMSSW_11_0_X for the HLT TDR.
This PR will not break the compatibility because it does not change the data format of the tracker digi. 
@suchandradutta could you confirm this PR will not break this compatibility?

suchandradutta commented:
Yes I can confirm that this PR will not break the compatibility

silviodonato commented:
Kind reminder of this PR to
upgrade:   @kpedro88 
simulation:   @mdhildreth  @civanch

kpedro88 commented:
+upgrade

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @silviodonato, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
